src/map/status.cpp

```diff
@@ -5841,13 +5841,28 @@ void status_calc_bl_main(struct block_list *bl, std::bitset<SCB_MAX> flag)
 	if(flag[SCB_SPEED]) {
 		status->speed = status_calc_speed(bl, sc, b_status->speed);
 
-		if( bl->type&BL_PC && !(sd && sd->state.permanent_speed) && status->speed < battle_config.max_walk_speed )
-			status->speed = battle_config.max_walk_speed;
+		if (bl->type & BL_PC && sd != nullptr) {
+			if (!sd->state.permanent_speed && status->speed < battle_config.max_walk_speed)
+				status->speed = battle_config.max_walk_speed;
 
+			// Recalculate homunculus speed if the player receives a speed buff/debuff
+			if (hom_is_active(sd->hd))
+				status_calc_bl(&sd->hd->bl, { SCB_SPEED, SCB_HOM });
+		}
 		if( bl->type&BL_PET && ((TBL_PET*)bl)->master)
 			status->speed = status_get_speed(&((TBL_PET*)bl)->master->bl);
-		if( bl->type&BL_HOM && battle_config.hom_setting&HOMSET_COPY_SPEED && ((TBL_HOM*)bl)->master)
-			status->speed = status_get_speed(&((TBL_HOM*)bl)->master->bl);
+		if( bl->type & BL_HOM) {
+			homun_data* hd = reinterpret_cast<homun_data*>(bl);
+
+			if (hd != nullptr && hd->master != nullptr) {
```

Comment:
```suggestion
			if (hd->master != nullptr) {
```
Cannot be null here.